src/twisted/web/http.py

```diff
@@ -377,6 +379,75 @@ def parseContentRange(header):
 
 
 
+class _IDeprecatedHTTPChannelToRequestInterface(Interface):
+    """
+    The interface L{HTTPChannel} expects of L{Request}.
+    """
+
+    requestHeaders = Attribute(
+        "A L{http_headers.Headers} instance giving all received HTTP request "
+        "headers.")
+
+    responseHeaders = Attribute(
+        "A L{http_headers.Headers} instance holding all HTTP response "
+        "headers to be sent.")
+
+
+    def connectionLost(reason):
+        """
+        The underlying connection has been lost.
+
+        @param reason: A failure instance indicating the reason why
+            the connection was lost.
+        @type reason: L{twisted.python.failure.Failure}
+        """
+
+
+    def gotLength(length):
+        """
+        Called when L{HTTPChannel} has determined the length, if any,
+        of the incoming request's body.
+
+        @param length: The length of the request's body.
+        @type length: L{int} if the request declares its body's length
+            and L{None} if it does not.
+        """
+
+
+    def handleContentChunk(data):
+        """
+        Deliver a received chunk of body data to the request.  Note
+        this does not imply chunked transfer encoding.
+
+        @param data: The received chunk.
+        @type data: L{bytes}
+        """
+
+
+    def parseCookies():
+        """
+        Parse the request's cookies out of received headers.
+        """
+
+
+    def requestReceived(command, path, version):
+        """
+        Called when the entire request, including its body, has been
+        received.
+
+        @param command: The request's HTTP command.
+        @type command: L{bytes}
+
+        @param path: The request's path.  Note: this is actually what
+            RFC7320 calls the URI.
+        @type path: L{bytes}
+
+        @param version: The request's HTTP version.
+        @type version: L{bytes}
+        """
+
+
```

Comment:
this is tricky enough that i'm willing to admit i'm stumped.  switching `!=` to `is not` [here](https://github.com/twisted/twisted/blob/e920939343ac5f6795c4509e0fe25c33a2447531/src/twisted/web/http.py#L1997) and removing the `__eq__`/`__ne__` implementation from the proxy breaks the tests, because [`Request._cleanup` passes the `Request` instance to `HTTPChannel.requestDone`](https://github.com/twisted/twisted/blob/e920939343ac5f6795c4509e0fe25c33a2447531/src/twisted/web/http.py#L632).  in other words, we end up with:

```python
if actualRequestInstance is not proxyAroundActualRequestInstance: raise TypeError
```

that's why the `__eq__` implementation does `self.original == other`.

the Completely Correct Approach, i think, would be to add `__ne__` to the interface and then make `Request` implement it.  that seemed pretty bad so i didn't do it, but maybe it's a good idea after all?